libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,831 @@
+#if HAL_WITH_ESC_TELEM
+        esc.set_state(ESCState::WANT_SEND_SET_TLM_TYPE);
+#else
+        esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+#endif
+        break;
+#endif
+
+#if HAL_WITH_ESC_TELEM
+    case ESCState::WANT_SEND_SET_TLM_TYPE:
+        return;
+    case ESCState::WAITING_SET_TLM_TYPE_OK:
+        if (buffer_contains_ok(length)) {
+            esc.set_state(ESCState::WANT_SEND_SET_FAST_COM_LENGTH);
+        }
+        break;
+#endif
+
+    case ESCState::WANT_SEND_SET_FAST_COM_LENGTH:
+        return;
+    case ESCState::WAITING_SET_FAST_COM_LENGTH_OK:
+        if (buffer_contains_ok(length)) {
+            esc.set_state(ESCState::RUNNING);
+        }
+        break;
+    case ESCState::RUNNING:
+        // we only expect telemetry messages in this state
+#if HAL_WITH_ESC_TELEM
+        if (!esc.telem_expected) {
+            // esc.unexpected_telem++;
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            AP_HAL::panic("unexpected telemetry");
+#endif
+            return;
+        }
+        esc.telem_expected = false;
+        return handle_message_telem(esc);
+#else
+        return;
+#endif
+
+    }
+}
+
+#if HAL_WITH_ESC_TELEM
+void AP_FETtecOneWire::handle_message_telem(ESC &esc)
+{
+    // the following two methods are coming from AP_ESC_Telem:
+    const TLM &tlm = u.packed_tlm.msg;
+
+    esc.error_count += tlm.tx_err_count;
+
+    // update rpm and error rate
+    float error_rate_pct = 0;
+    if (_sent_msg_count) {
+        error_rate_pct = esc.error_count/(float)_sent_msg_count;
+    }
+    update_rpm(esc.servo_ofs,
+               tlm.rpm*100*2/_pole_count_parameter,
+               error_rate_pct);
+
+    // update power and temperature telem data
+    TelemetryData t {};
+    t.temperature_cdeg = tlm.temp * 100;
+    t.voltage = tlm.voltage * 0.01f;
+    t.current = tlm.voltage * 0.01f;
+    t.consumption_mah = tlm.consumption_mah;
+    update_telem_data(
+        esc.servo_ofs,
+        t,
+        TelemetryType::TEMPERATURE|
+          TelemetryType::VOLTAGE|
+          TelemetryType::CURRENT|
+          TelemetryType::CONSUMPTION);
+
+    esc.last_telem_us = AP_HAL::micros64();
+}
+#endif
+
+// reads data from the UART, calling handle_message on any message found
+void AP_FETtecOneWire::read_data_from_uart()
+{
+    /*
+    a frame looks like:
+    byte 1 = frame header (0x02 = bootloader, 0x03 = ESC firmware)
+    byte 2 = sender ID (5bit)
+    byte 3 & 4 = frame type (always 0x00, 0x00 used for bootloader. here just for compatibility)
+    byte 5 = frame length over all bytes
+    byte 6 - X = answer type, followed by the payload
+    byte X+1 = 8bit CRC
+    */
+
+#if HAL_AP_FETTEC_HALF_DUPLEX
+    //ignore own bytes
+    if (_use_hdplex) {
+        while (_ignore_own_bytes > 0 && _uart->available()) {
+            _ignore_own_bytes--;
+            _uart->read();
+        }
+    }
+#endif
+
+    uint32_t bytes_to_read = MIN(_uart->available(), 1024U);

```

Comment:
Why such an high limit? Why not just 128? A telemetry packet is at most 17bytes long.

Reply:
128 is fine.  This number is just here so we don't spend a huge amount of time processing serial data in the loop.

Reply:
done